apollo-router/src/plugins/telemetry/tracing/apollo_telemetry.rs

```diff
@@ -135,23 +176,50 @@ pub(crate) enum Error {
     SystemTime(#[from] SystemTimeError),
 }
 
+// TBD(tim): explore ways to have different types of LightSpanData,
+// since we only need some of these extra fields if we are going to send to OTel.
 #[derive(Debug, Clone, PartialEq)]
-struct LightSpanData {
-    span_id: SpanId,
-    name: Cow<'static, str>,
-    start_time: SystemTime,
-    end_time: SystemTime,
-    attributes: EvictedHashMap,
+pub(crate) struct LightSpanData {
+    pub(crate) trace_id: TraceId,
+    pub(crate) span_id: SpanId,
+    pub(crate) parent_span_id: SpanId,
+    pub(crate) span_kind: SpanKind,
+    pub(crate) name: Cow<'static, str>,
+    pub(crate) start_time: SystemTime,
+    pub(crate) end_time: SystemTime,
+    pub(crate) attributes: EvictedHashMap,
 }
 
-impl From<SpanData> for LightSpanData {
-    fn from(value: SpanData) -> Self {
+impl LightSpanData {
+    /// Convert from a full Span into a lighter more memory-efficient span for caching purposes.
+    /// - If `include_attr_names` is passed, filter out any attributes that are not in the list.
+    fn from_span_data(value: SpanData, include_attr_names: &Option<HashSet<Key>>) -> Self {
+        let filtered_attributes = match include_attr_names {
+            None => value.attributes,
+            Some(attr_names) => {
+                // Looks like this transformation will be easier after upgrading opentelemetry_sdk >= 0.21
+                // when attributes are stored as Vec<KeyValue>.
+                // https://github.com/open-telemetry/opentelemetry-rust/blob/943bb7a03f9cd17a0b6b53c2eb12acf77764c122/opentelemetry-sdk/CHANGELOG.md?plain=1#L157-L159
+                let max_attr_len = std::cmp::min(attr_names.len(), value.attributes.len());
+                let mut new_attrs = EvictedHashMap::new(max_attr_len.try_into().unwrap(), max_attr_len);
+                value.attributes.iter().for_each(|(key, value)| {
```

Comment:
If you want to avoid copies then use `into_iter()` instead and then in the loop you'll be able to delete the clone
```suggestion
                value.attributes.into_iter().for_each(|(key, value)| {
```